src/enclosure.c

```diff
@@ -171,3 +184,154 @@ int enclosure_open(const struct enclosure_device *enclosure)
 
 	return fd;
 }
+
+static struct enclosure_device *find_enclosure(char *enclosure_id)
+{
+	struct enclosure_device *enclosure_device;
+	list_for_each(sysfs_get_enclosure_devices(), enclosure_device) {
+		if (strncmp(enclosure_device->dev_path, enclosure_id, PATH_MAX) == 0) {
+			return enclosure_device;
+		}
+	}
+	return NULL;
+}
+
+status_t find_block_device_by_enclosure_slot(struct enclosure_device *encl, int index, enum ibpi_pattern *status, struct block_device **device)
+{
+	*status = IBPI_PATTERN_NORMAL;
+	*device = NULL;
+	for (int i = 0; i < encl->slots_count; i++) {
+		if (encl->slots[i].index == index) {
+			*status = encl->slots[i].ibpi_status;
+			*device = locate_block_by_sas_addr(encl->slots[i].sas_addr);
+			return STATUS_SUCCESS;
+		}
+	}
+	return STATUS_DATA_ERROR;
+}
+
+status_t enclosure_get_slot_by_device(char *device, struct slot_response *slot_res)
+{
+	struct block_device *block_device = get_block_device_from_sysfs_path(basename(device));
+
+	if (!block_device) {
+		log_error("SCSI: Device node not found %s\n", device);
+		return STATUS_INVALID_PATH;
+	}
+
+	if (!block_device->enclosure) {
+		log_error("SCSI: Not a SCSI ses device %n\n", device);
+		return STATUS_INVALID_PATH;
+	}
+
+	snprintf(slot_res->slot, PATH_MAX, "%s/%d", block_device->enclosure->dev_path, block_device->encl_index);
+	strcpy(slot_res->device, device);
```

Comment:
Please use snprintf here. We will run static code analysis on your change and I'm sure that here will be error because in some extraordinary cases '\0' at end can be missed.

Reply:
Done